DM-47770: Fix threadsafety of sqlalchemy MetaData access

dhirving opened the pull request:
Wrap all accesses to sqlalchemy.MetaData with a lock to avoid concurrency issues.

sqlalchemy.MetaData is documented to be threadsafe for reads, but not with concurrent modifications.  We add tables dynamically at runtime, and the MetaData object is shared by all Database instances sharing the same connection pool.

Prior to adding the lock, Butler server database calls that added table definitions dynamically were sometimes failing with InvalidRequestError exceptions complaining about inconsistency of table definitions.

## Checklist

- [x] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`
- [ ] (if changing dimensions.yaml) make a copy of dimensions.yaml in `configs/old_dimensions`

dhirving committed:
Fix threadsafety of sqlalchemy MetaData access

Wrap all accesses to sqlalchemy.MetaData with a lock to avoid concurrency issues.

sqlalchemy.MetaData is documented to be threadsafe for reads, but not with concurrent modifications.  We add tables dynamically at runtime, and the MetaData object is shared by all Database instances sharing the same connection pool.

Prior to adding the lock, Butler server database calls that added table definitions dynamically were sometimes failing with InvalidRequestError exceptions complaining about inconsistency of table definitions.

dhirving committed:
Fix failure of sqlite test when using system temp dir

Fix an issue where a test for equality of filenames in the system temp directory were failing.  On MacOS, /private/tmp and /tmp refer to the same directory.

andy-slac commented on python/lsst/daf/butler/registry/interfaces/_database.py:
```diff
@@ -164,13 +165,9 @@ def addTable(self, name: str, spec: ddl.TableSpec) -> sqlalchemy.schema.Table:
         to be declared in any order even in the presence of foreign key
         relationships.
         """
-        name = self._db._mangleTableName(name)
```
I am confused, do you want `DatabaseMetadata` to handle table name mangling internally? I do not think it is done consistently now, you call `add_table` with the original name, but `get_table` is called with mangled name?

dhirving replied:
I am just going to remove it entirely.  It's a no-op -- the only implementation returns the name unchanged.  I think it is a holdover from the Oracle implementation.  

The existing implementation was already incorrect because there were several places where name mangling was applied multiple times.  I noticed that the two cases I already removed were double-applying the mangling (because it is done internally in _convertTableSpec) so I had removed those, but apparently the logic is also wrong in a lot of other places.

TallJimbo replied:
My recollection is that it was PostgreSQL, not Oracle, that needed name shrinking.  It may have been index and constraint names more than table names where it was important, but I'm pretty sure PostgreSQL is the one with a 64-char limit.

dhirving replied:
Two different things.  The name shrinking logic is still there.

TallJimbo replied:
Ah, sorry for the noise.

andy-slac commented on python/lsst/daf/butler/registry/interfaces/_database.py:
```diff
@@ -1192,7 +1189,7 @@ def getExistingTable(self, name: str, spec: ddl.TableSpec) -> sqlalchemy.schema.
         """
         assert self._metadata is not None, "Static tables must be declared before dynamic tables."
         name = self._mangleTableName(name)
```
Maybe this one needs to be dropped too?

andy-slac commented on python/lsst/daf/butler/registry/interfaces/_database.py:
```diff
@@ -1206,10 +1203,7 @@ def getExistingTable(self, name: str, spec: ddl.TableSpec) -> sqlalchemy.schema.
             )
             if name in inspector.get_table_names(schema=self.namespace):
```
Hmm, but this won't work if name is not mangled?

andy-slac reviewed (approved):
Looks good in general, but I'm not sure that name mangling is done consistently.

dhirving committed:
Remove defunct table name mangling code

This table name mangling logic is a holdover from the Oracle implementation and the current implementation is a no-op.  The logic was being inconsistently applied so none of this would have worked even in a new implementation that had a real implementation of the mangling function.
